dcicutils/structured_data.py

```diff
@@ -0,0 +1,792 @@
+import sys
+from typing import Any, Callable, List, Optional, Tuple, Type, Union
+from webtest.app import TestApp, TestResponse
+from dcicutils.common import OrchestratedApp, APP_CGAP, APP_FOURFRONT, APP_SMAHT, ORCHESTRATED_APPS
+from dcicutils.creds_utils import CGAPKeyManager, FourfrontKeyManager, SMaHTKeyManager
+from dcicutils.data_readers import CsvReader, Excel, RowReader
+from dcicutils.ff_utils import get_metadata, get_schema, patch_metadata, post_metadata
+from dcicutils.misc_utils import (create_object, load_json_if, merge_objects, remove_empty_properties, right_trim,
+                                  split_string, to_boolean, to_camel_case, to_enum, to_float, to_integer, VirtualApp)
+from dcicutils.zip_utils import temporary_file, unpack_gz_file_to_temporary_file, unpack_files
+
+
+# Classes/functions to parse a CSV or Excel Spreadsheet into structured data, using a specialized
+# syntax to allow structured object properties to be referenced by column specifiers. This syntax
+# uses an (intuitive) dot notation to reference nested objects, and a (less intuitive) notation
+# utilizing the "#" character to reference array elements. May also further coerce data types by
+# consulting an optionally specified JSON schema.
+#
+# Alternate and semantically equivalent implementation of dcicutils.{sheet,bundle}_utils.
+# Spare time exercise, with benefit of sheet_utils implementation experience.
+
+ACCEPTABLE_FILE_SUFFIXES = [".csv", ".tsv", ".json", ".xls", ".xlsx", ".gz", ".tar", ".tar.gz", ".tgz", ".zip"]
+ARRAY_VALUE_DELIMITER_CHAR = "|"
+ARRAY_VALUE_DELIMITER_ESCAPE_CHAR = "\\"
+ARRAY_NAME_SUFFIX_CHAR = "#"
+ARRAY_NAME_SUFFIX_REGEX = re.compile(rf"{ARRAY_NAME_SUFFIX_CHAR}\d+")
+DOTTED_NAME_DELIMITER_CHAR = "."
+
+# Forward type references for type hints.
+Portal = Type["Portal"]
+PortalBase = Type["PortalBase"]
+Schema = Type["Schema"]
+StructuredDataSet = Type["StructuredDataSet"]
+
+
+class StructuredDataSet:
+
+    def __init__(self, file: Optional[str] = None, portal: Optional[Union[VirtualApp, TestApp, Portal]] = None,
+                 schemas: Optional[List[dict]] = None, data: Optional[List[dict]] = None,
+                 order: Optional[List[str]] = None, prune: bool = True) -> None:
+        self.data = {} if not data else data  # If portal is None then no schemas nor refs.
+        self._portal = Portal(portal, data=self.data, schemas=schemas) if portal else None
+        self._order = order
+        self._prune = prune
+        self._issues = {}
+        self._refs_resolved = []
+        self._validated = False
+        self._load_file(file) if file else None
+
+    @staticmethod
+    def load(file: str, portal: Optional[Union[VirtualApp, TestApp, Portal]] = None,
+             schemas: Optional[List[dict]] = None,
+             order: Optional[List[str]] = None, prune: bool = True) -> StructuredDataSet:
+        return StructuredDataSet(file=file, portal=portal, schemas=schemas, order=order, prune=prune)
```

Comment:
Some docstring may be useful especially in situations where you have a lot of complex/mixed types as input